tasks/retrieval_deal.go

```diff
@@ -37,30 +69,98 @@ func MakeRetrievalDeal(ctx context.Context, config NodeConfig, node api.FullNode
+	de.log("got query offer", "root", offer.Root, "piece", offer.Piece, "size", offer.Size, "minprice", offer.MinPrice, "unseal_price", offer.UnsealPrice)
+	return nil
+}
+
+func (de *retrievalDealExecutor) executeAndMonitorDeal(updateStage UpdateStage) error {
+	dealStage := RetrievalStages["ProposeRetrieval"]
+	err := updateStage("ProposeRetrieval", &dealStage)
+	if err != nil {
+		return err
+	}
 
 	ref := &api.FileRef{
-		Path:  filepath.Join(config.NodeDataDir, "ret"),
-		IsCAR: task.CARExport,
+		Path:  filepath.Join(de.config.NodeDataDir, "ret"),
+		IsCAR: de.task.CARExport,
 	}
 
-	err = node.ClientRetrieve(ctx, offer.Order(config.WalletAddress), ref)
+	events, err := de.node.ClientRetrieveWithEvents(de.ctx, de.offer.Order(de.config.WalletAddress), ref)
 	if err != nil {
 		return err
 	}
 
-	rdata, err := ioutil.ReadFile(filepath.Join(config.DataDir, "ret"))
+	AddLog(&dealStage, "deal sent to miner")
+	err = updateStage("ProposeRetrieval", &dealStage)
 	if err != nil {
 		return err
 	}
 
-	log("retrieval successful", "PayloadCID", task.PayloadCID)
+	lastStatus := retrievalmarket.DealStatusNew
+	var lastBytesReceived uint64 = 0
+	for event := range events {
+		if event.Status != lastStatus {
+			de.log("Deal status",
+				"cid", de.task.PayloadCID,
+				"state", retrievalmarket.DealStatuses[event.Status],
+				"error", event.Err,
+				"received", event.BytesReceived,
+			)
+			lastStatus = event.Status
+		}
+
+		if event.Event == retrievalmarket.ClientEventDealAccepted {
+			dealStage = RetrievalStages["DealAccepted"]
+			err := updateStage("DealAccepted", &dealStage)
+			if err != nil {
+				return err
+			}
+		}
+		if event.BytesReceived > 0 && lastBytesReceived == 0 {
+			dealStage = RetrievalStages["FirstByteReceived"]
+			err := updateStage("FirstByteReceived", &dealStage)
+			if err != nil {
+				return err
+			}
+		}
+		switch event.Status {
+		case retrievalmarket.DealStatusCancelled,
+			retrievalmarket.DealStatusErrored,
+			retrievalmarket.DealStatusRejected,
+			retrievalmarket.DealStatusDealNotFound:
+
+			return errors.New("storage deal failed")
 
-	_ = rdata
+		// deal is on chain, exit successfully
+		case retrievalmarket.DealStatusCompleted:
+			dealStage = RetrievalStages["DealComplete"]
+			AddLog(&dealStage, fmt.Sprintf("bytes received: %d", event.BytesReceived))
+			err := updateStage("DealComplete", &dealStage)
+			if err != nil {
+				return err
+			}
 
-	//if carExport {
-	//rdata = ExtractCarData(ctx, rdata, rpath)
-	//}
+			rdata, err := ioutil.ReadFile(filepath.Join(de.config.DataDir, "ret"))
+			if err != nil {
+				return err
+			}
+
+			de.log("retrieval successful", "PayloadCID", de.task.PayloadCID)
+
+			_ = rdata
+
+			AddLog(&dealStage, "file read from file system")
+			err = updateStage("DealComplete", &dealStage)
+			if err != nil {
+				return err
+			}
+
+			//if carExport {
+			//rdata = ExtractCarData(ctx, rdata, rpath)
+			//}
```

Comment:
Should probably remove this, or make it return `ErrNotImplemented`.

Reply:
yea agree it's a copy paste but will change